### ISSUE TITLE
Uneeded "Format" on Logout Request

### ISSUE THREAD

**Issue by Blueowner** [onelogin/php-saml#205]
Hello,
I'm trying to debug an issue with the logout process.

After a couple hours of debugging I found out that commenting out this line: https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Utils.php#L959 would make things work correctly. I can post the log I get from SAML Tracer (Firefox extension) if needed.

I'm no expert in AD FS nor Saml2, so before going for a Pull Request that might not make any sense, I've a quick question:
Shouldn't the line https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Utils.php#L959, `$nameId->setAttribute('Format', $format);` be inside the `if (isset($spnq)) {` three lines above?

From the function's DocBlock it kinda makes sense to me that the SP Format should only be present if the SP Name Qualifier is set.

Am I missing something?

PS. I'm using php-saml with https://github.com/aacotroneo/laravel-saml2 so it might be "laravel-saml2" that is not using php-saml correctly.

**Comment by Blueowner** [comment]
After some more debugging I found out that it might be a problem with checking whether `$nameIdFormat` is `empty()` or `is_null()`. I think that `$nameIdFormat = false` should not be threaded the same way as `$nameIdFormat = null` since a `false` value will generate a different XML than `null`.

That's the two edits I've done to get it working:
- https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Auth.php#L450 from
`public function logout($returnTo = null, $parameters = array(), $nameId = null, $sessionIndex = null, $stay = false, $nameIdFormat = null)` to
`public function logout($returnTo = null, $parameters = array(), $nameId = null, $sessionIndex = null, $stay = false, $nameIdFormat = false)`

- https://github.com/onelogin/php-saml/blob/master/lib/Saml2/LogoutRequest.php#L68 from
`if (empty($nameIdFormat)) {` to
`if (is_null($nameIdFormat)) {`.

If you think this is an issue, I can create a pull request.

**Comment by pitbulk** [comment]
[Here](http://www.datypic.com/sc/saml2/e-saml_NameID.html) you can see how the NameID elemnt is built. The Format makes reference to the format of the NameID's value.

The constructor will try to set the Format provided in the method, if not it will try to set the format that appears on the [settings](https://github.com/onelogin/php-saml/blob/master/settings_example.php#L62), by default set to [UNSPECIFIED](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Settings.php#L338)

PHP empty method will return true if the provided format is null, false or "" , so if that is the case, the default settings will be used.

But I don't understand why provide false as nameIdFormat.

[lavarel-saml2](https://github.com/aacotroneo/laravel-saml2/blob/master/src/Aacotroneo/Saml2/Saml2Auth.php#L66) is not providing any value for the nameIdFormat so its default setting value should be used, in this case ["persistent"](https://github.com/aacotroneo/laravel-saml2/blob/master/src/config/saml2_settings.php#L81) and maybe that value is causing you problems.

**Comment by Blueowner** [comment]
The problem I'm facing is "The SAML Single Logout request does not correspond to the logged-in session participant." (from the Windows Event Viewer).

I think the error is thrown because the request doesn't contain the NameIdFormat (nor the SPNameQualifier). I've been able to generate a request without the "SPNameQualifier" but not without the "NameIdFormat".

I've trying to modify the login request to include the "NameIdFormat" but I was not able to, probably because of a lack of knowledge of AD FS and Saml2.

I've created this morning a [Pull request](https://github.com/aacotroneo/laravel-saml2/pull/81) on the laravel-saml2 repository to allow to pass a value for "nameIdFormat".

I can post the saml2 settings and the Event Viewer logs if needed.

**Comment by pitbulk** [comment]
Close since that issue belong laravel-saml2

**Comment by Blueowner** [comment]
Hello,
sorry for the late feedback.
I've reviewed your indications this morning for a second time.

> But I don't think why a developer should provide false as nameIdFormat value on the @ constructor...

I've found the need to have a SAML2 logout request without the "NameIdFormat" property, it's the only way I got the logout working.

As of my understanding, the XML of the login and logout request has to have the same parameters. Given the error I get when logging out through AD FS, I came to the conclusion that to solve my problem I either have to add the "NameIdFormat" parameter when performing the login request, something I haven't been able to do, or make sure the logout request doesn't have the "NameIdFormat" property.

Even if laravel-saml2 merges [this pull request](https://github.com/aacotroneo/laravel-saml2/pull/81), it would still be necessary to have the change I've proposed in one of my previous comments implemented into the php-saml library. We wouldn't be able to generate the appropriate XML otherwise.

I think the following still applies and I would love to get your opinion on this:

> After some more debugging I found out that it might be a problem with checking whether `$nameIdFormat` is `empty()` or `is_null()`. I think that `$nameIdFormat = false` should not be threaded the same way as `$nameIdFormat = null` since a `false` value will generate a different XML than `null`.
